File: Easy/Arrays_Strings/125_Valid_Palindrome.c
/*
Problem: Valid Palindrome

Approach:
- Traverse the string and retain only alphanumeric characters while converting them to lowercase

Time Complexity: O(n)
Space Complexity: O(1)
*/

#include <string.h>
#include <stdbool.h>

bool isPalindrome(char* s) {
    int i = 0, j = 0;
    // Remove non alpha-numeric chars
    while (s[i]) {
        if (isalnum(s[i])) {
            s[j++] = tolower(s[i]); // Convert to lowercase
        }
        i++;
    }
    s[j] = '\0';   // Terminate the string

    int length = j;
    // Check palindrome
    for (i = 0; i < length / 2; i++) {
        if (s[i] != s[length - i - 1]) {
            return false;
        }
    }
    return true;
}

File: Easy/Arrays_Strings/125_Valid_Palindrome.py
'''
Problem: Valid Palindrome

Approach:
- Traverse the string and retain only alphanumeric characters while converting them to lowercase
- Use 2 pointers to check if all characters are same from left and right sides

Time Complexity: O(n)
Space Complexity: O(1)
'''

class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = ''.join(char.lower() for char in s if char.isalnum())
        
        left = 0
        right = len(s) - 1
        for i in range(len(s)//2):
            if s[left] != s[right]:
                return False
                
            left+=1
            right-=1
        
        return True
